### PR TITLE
Improve multi-application experience, allow including suites by name (opposite of --skip)

### PR DESCRIPTION
**Pull request by calvinalkan**
Fix: #6434

This PR is a minimal solution to fix/support the issue described in #6434.

This allows to run suites by name for included applications-

Test cases are included.

### CODE REVIEW

**Comment by calvinalkan** [comment]
I don't know why some remote coverage tests would fail for 7.2 only, nothing related has been changed there.

**Comment by DavertMik** [comment]
@calvinalkan thanks for the PR
I got the problem but I don't like the proposed solution and extra options added to CLI.

How about adding more verbose suite syntax instead, for instance 

* `codecept run unit` <- executes unit tests for current app
* `codecept run backed:unit` <- executes unit tests for backend app
* `codecept run *:unit` <- executes unit tests for app included apps (your case, as I understood)


This is just an idea but it is a more general solution and probably it can be implemented in a more beautiful way.

**Comment by calvinalkan** [comment]
@DavertMik Thanks for the comment.

I agree the options are not ideal but this was the least change required solution.

`codecept run backed:unit` - What exactly should `backend` be here?

The relative path (from cwd) to a `codeception.yml` file I assume.

I'm using this in the context of a very big monorepo, so it would look like this for me:

`codecept run src/packages/package-a:unit`
`codecept run src/packages/package-b:unit`
`codecept run src/packages/package-a:functional`
`codecept run src/packages/package-b:functional`

Now running all tests for all packages by suite.

`codecept run *:unit`
`codecept run *:functional`

I like this solution a lot more than what I proposed.

Maybe we could even do:

`codecept run src/packages/package-a unit` (notice the space)

But this probably clashes a lot with how it currently works.

This changes should probably be done in the main method of the `Run` class right?

**Comment by calvinalkan** [comment]
The only question is how should we run:

`codecept run -i unit -i functional`

Options:

1. `codecept run *:unit:functional` 
2. ...?

**Comment by DavertMik** [comment]
@calvinalkan If you want to run 2 suites it would be 

```
codecept run *:unit,*:functional
```

as we have similar comma syntax for suites themselves:

```
codecept run unit,functioal
```

Sorry for delay in reply

**Comment by calvinalkan** [comment]
@DavertMik Based on your feedback I implemented the following:

Runs unit and functional suites for all included apps
`
codecept run *::unit,*::functional
`

Runs unit and functional test for backend app
`codecept run backend::unit,backed::functional`

Same as above, but with nested package directory
`codecept run /path/to/nested/package::unit`

I chose using `::` instead of `:` because `:` is already used for filtering. E.g (`php vendor/bin/codecept run tests/acceptance/backend:^login`)

**Comment by DavertMik** [comment]
Good idea, thank you!
That will be awesome feature.

**Comment by calvinalkan** [comment]
@DavertMik Any idea what's going on on PHP <= 7.1?

**Comment by DavertMik** [comment]
Please make pull request to 5.0 branch which drops php 7 support

**Comment by calvinalkan** [comment]
I think I found the issue. On PHP7.1 it uses a prehistoric version of `symfony/console` which does not yet have the exit code constants.

This should work now.

I'd love to get this in 4.x as we are bound to it due to `wp-browser` not being there yet https://github.com/lucatume/wp-browser/issues/571

**Comment by calvinalkan** [comment]
@DavertMik All working now

File: src/Codeception/Command/Run.php [hunk 1]
```diff
@@ -405,18 +408,60 @@ public function execute(InputInterface $input, OutputInterface $output)
...
         // Run all tests of given suite or all suites
         if (!$test) {
-            $suites = $suite ? explode(',', $suite) : Configuration::suites();
-            $this->executed = $this->runSuites($suites, $this->options['skip']);
-
-            if (!empty($config['include']) and !$suite) {
+            $raw_suites = $suite ? explode(',', $suite) : Configuration::suites();
+            
+            /** @var string[] $main_app_suites */
+            $main_app_suites = [];
```
Comment (Naktibalda) [inline src/Codeception/Command/Run.php:414]: Please rename all your variables to snakeCase, e.g. `$mainAppSuites`.

**Review by Naktibalda — CHANGES_REQUESTED** [review]

File: src/Codeception/Command/Run.php (hunk 1 above)
Comment (calvinalkan) [reply src/Codeception/Command/Run.php]: @Naktibalda done

File: src/Codeception/Command/Run.php [hunk 2]
```diff
@@ -488,8 +533,10 @@ protected function matchSingleTest($suite, $config)
      *
      * @param array $suites
      * @param string $parent_dir
+     * @param array<string,string[]> $filter_app_suites An array keyed by included app name where values are suite names to run.
+     * @param string[] $filter_suites_by_wildcard A list of suite names (applies to all included apps)
      */
-    protected function runIncludedSuites($suites, $parent_dir)
+    protected function runIncludedSuites($suites, $parent_dir, $filter_app_suites = [], $filter_suites_by_wildcard = [])
```
Comment (Naktibalda) [inline src/Codeception/Command/Run.php:539]: Please rename variables in this method too.

File: tests/data/snapshots/tests/_data/Snapshot.NotAJsonSnapshot.xml [hunk 3]
```diff
@@ -1,8 +1,8 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema"></xs:schema>
 <note>
-    <to>Son</to>
-    <from>Vader</from>
-    <heading>Disclaimer</heading>
-    <body>I'm your father!</body>
+    <to>Tove</to>
+    <from>Jani</from>
```
Comment (Naktibalda) [inline tests/data/snapshots/tests/_data/Snapshot.NotAJsonSnapshot.xml:5]: Please revert changes of this file.

One day I will fix the test which makes this change :)

**Comment by Naktibalda** [comment]
@DavertMik This change is a feature. Should we bump version number to 4.2.0?

**Review by Naktibalda — CHANGES_REQUESTED** [review]

**Comment by calvinalkan** [comment]
> @DavertMik This change is a feature. Should we bump version number to 4.2.0?

That would be great.

**Comment by calvinalkan** [comment]
Don't merge this yet. I found an issue that has no tests currently.

Or maybe it's not an issue but what about this scenario (might be an edge case):


What it the root codeception.yml file has some tests and includes other apps:

```yml
include:
  - src/Snicco/Component/*
  - src/Snicco/Bundle/*
paths:
  output: tests/_output
  tests: tests
  data: tests/_data
  support: tests/_support
settings:
  colors: true
bootstrap: bootstrap.php
reporters:
  report: "PhpStorm_Codeception_ReportPrinter"
```

This would be for example the case with a monorepo where you have some root level tests for utilities related to the monorepo and each included app has its own separate test suites.

On my fork currently when you run (with the config above)

`codecept run unit` the following happens:

- the unit suite from the root repo is run, all other suites from included apps are run. This is not desired of course and I think we can prevent this by just adding a check for a suite [here](https://github.com/Codeception/Codeception/blob/30b102e2a343e5ee6583a00b6923b2398fd2a068/src/Codeception/Command/Run.php#L439).

Ill fix that now, 

The last question is:

When I run `codecept run *::unit`, should that include the unit suite of the root application (if present). I would say yes.

**Comment by calvinalkan** [comment]
@DavertMik @Naktibalda 

I added some new tests that now verify previously uncovered behavior:

Tests are pretty self-explanatory but in resume:


 - `codecept run` => Everything is run
 - `codecept run unit` => Runs unit suite from current app
 - `codecept run *::unit` => Runs unit suites from all included apps and **NOT** the root suite
 - `codecept run unit,*::unit` => Runs included unit suites **AND** root unit suite 
 - `codecept run functional,*::unit` => Runs included unit suites and root **functional** suite
 
 I think this is the most flexible approach.

**Comment by Naktibalda** [comment]
Please rebase your test on 4.2 branch and fix merge conflict in IncludedCest.

File: src/Codeception/Command/Run.php [hunk 4]
```diff
@@ -488,14 +541,16 @@ protected function matchSingleTest($suite, $config)
...
      */
-    protected function runIncludedSuites($suites, $parent_dir)
+    protected function runIncludedSuites($suites, $parent_dir, $filterAppSuites = [], $filterSuitesByWildcard = [])
     {
         $defaultConfig = Configuration::config();
         $absolutePath = \Codeception\Configuration::projectDir();
 
         foreach ($suites as $relativePath) {
-            $current_dir = rtrim($parent_dir, DIRECTORY_SEPARATOR) . DIRECTORY_SEPARATOR . $relativePath;
+            $current_dir = rtrim($parent_dir, DIRECTORY_SEPARATOR).DIRECTORY_SEPARATOR.$relativePath;
```
Comment (Naktibalda) [inline src/Codeception/Command/Run.php:553]: Please restore spaces around ` . ` in concatenation.

File: src/Codeception/Command/Run.php [hunk 5]
```diff
@@ -506,7 +561,15 @@ protected function runIncludedSuites($suites, $parent_dir)
             }
 
             $suites = Configuration::suites();
-
+            
+            if(!empty($filterSuitesByWildcard)){
+                $suites = array_intersect($suites, $filterSuitesByWildcard);
+            }
+            
+            if(isset($filterAppSuites[$relativePath])) {
```
Comment (Naktibalda) [inline src/Codeception/Command/Run.php:569]: And add space after `if`   - `if (`

File: src/Codeception/Command/Run.php [hunk 6]
```diff
@@ -330,7 +330,7 @@ public function execute(InputInterface $input, OutputInterface $output)
 
         $suite = $input->getArgument('suite');
         $test = $input->getArgument('test');
-
+
```
Comment (Naktibalda) [inline src/Codeception/Command/Run.php:333]: Please remove spaces here.

File: src/Codeception/Command/Run.php [hunk 7]
```diff
@@ -385,10 +386,14 @@ public function execute(InputInterface $input, OutputInterface $output)
                 }
             }
         }
-
+
```
Comment (Naktibalda) [inline src/Codeception/Command/Run.php:389]: and here.

**Review by Naktibalda — CHANGES_REQUESTED** [review]
Code style issues

File: src/Codeception/Command/Run.php (hunk 7 above)
Comment (calvinalkan) [reply src/Codeception/Command/Run.php:389]: Ok, will be done within the next days

**Comment by calvinalkan** [comment]
@Naktibalda Rebased onto 4.2

**Review by DavertMik — APPROVED** [review]
Great job!
Looks good to me!

**Comment by calvinalkan** [comment]
@DavertMik Now we only have to fix running failed tests in multi-app and then it's pretty awesome for monorepos.

File: src/Codeception/Command/Run.php [hunk 8]
```diff
@@ -3,8 +3,6 @@
 
 use Codeception\Codecept;
 use Codeception\Configuration;
-use Codeception\Lib\GroupManager;
-use Codeception\Util\PathResolver;
 use Symfony\Component\Console\Command\Command;
 use Symfony\Component\Console\Input\InputArgument;
 use Symfony\Component\Console\Input\InputInterface;
```
Comment (Naktibalda) [inline src/Codeception/Command/Run.php:8]: Please document new syntax in `Usage` section of class level docblock in this file.  (Line 24)

**Review by Naktibalda — CHANGES_REQUESTED** [review]

**Comment by Naktibalda** [comment]
> @DavertMik Now we only have to fix running failed tests in multi-app and then it's pretty awesome for monorepos.

Does this change break some tests?
